packages/state-machine/test/index.test.ts

```diff
@@ -0,0 +1,162 @@
+import { describe, test, expect, vi } from "vitest";
+import { createMemo, createRoot, onCleanup, untrack } from "solid-js";
+import { createMachine } from "../src";
+
+describe("createMachine", () => {
+  test("switches state", () =>
+    createRoot(() => {
+      const state = createMachine<{
+        idle: { value: "foo" };
+        loading: { value: "bar" };
+      }>({
+        initial: "idle",
+        states: {
+          idle: () => "foo",
+          loading: () => "bar",
+        },
+      });
+      const memo = createMemo(() => {
+        // track only the signal
+        const v = state();
+        return untrack(() => ({ ...v }));
+      });
+
+      expect(state.type).toBe("idle");
+      expect(state.value).toBe("foo");
+      expect(state.type).toBe(memo().type);
+      expect(state.value).toBe(memo().value);
+
+      // @ts-expect-error need to check if state is idle
+      state.to.loading();
+
+      expect(state.type).toBe("loading");
+      expect(state.value).toBe("bar");
+      expect(state.type).toBe(memo().type);
+      expect(state.value).toBe(memo().value);
+
+      if (state.type === "loading") {
+        // @ts-expect-error can't go from loading to loading
+        state.to.loading();
+
+        state.to.idle();
+      }
+
+      expect(state.type).toBe("idle");
+      expect(state.value).toBe("foo");
+      expect(state.type).toBe(memo().type);
+      expect(state.value).toBe(memo().value);
+    }));
+
+  test("switches state with input", () => {
+    createRoot(() => {
+      const state = createMachine<{
+        idle: { input: string; value: any };
+        loading: { input: number; value: any };
+      }>({
+        initial: {
+          type: "idle",
+          input: "foo",
+        },
+        states: {
+          idle: i => i,
+          loading: i => i,
+        },
+      });
+
+      expect(state.type).toBe("idle");
+      expect(state.value).toBe("foo");
+
+      if (state.type === "idle") {
```

Comment:
This if is superfluous. The test should have already thrown otherwise.

Reply:
The if is for types. as you can see in test above that it won’t let you switch to different state before checking if that’s possible. So you have to narrow down the current state.

Reply:
Ah, too bad that typescript can't infer the current state from previous statements.

Reply:
yeah, and I'm wondering if the getters on the signal should be removed too, as type narrowing can be misleading there. With a function it's more clear that you have to call it again to get the new state.

![image](https://github.com/solidjs-community/solid-primitives/assets/24491503/d62351a9-633b-4a48-b468-07134a3439c5)

But then typechecking on `<Show>`, `<Switch>`, etc. becomes annoying

Like with getters you can do this
```ts
<Show when={state.type === "Reading" && state}>
  {narrowed_state => {
    narrowed_state // state for "Reading"
  }}
</Show>
```

With function it's more like
```ts
<Show when={(() => {
  const v = state()
  return v.type === "Reading" && v
})()}>
  {narrowed_state => {
    narrowed_state // state for "Reading"
  }}
</Show>
```

Reply:
What about the opposite - just getters, no memo/signal to call at all?

Reply:
The signal form is more idiomatic, it makes it obvious what changes, because in the end it is just a single signal. The getters are just an additional convenience, they are not separate signals, and they shouldn't be by default.
It also makes the usage with primitives expecting a function source easier: (`on`, `createSelector`, `createResource`, etc)
```ts
const isNodeDragged = createSelector(
  state,
  (node: Node, v) => v.type === "Dragging" && v.value.node === node,
)
// vs
const isNodeDragged = createSelector(
  () => ({ ...state }),
  (node: Node, v) => v.type === "Dragging" && v.value.node === node,
)
```